lib/net.c

```diff
@@ -1200,12 +1201,119 @@ start_iface(struct gatekeeper_if *iface, unsigned int num_attempts_link_get)
 	return ret;
 }
 
+static void
+calculate_net_config_para(struct net_config *net_conf)
+{
+	int i;
+
+	/* The total number of RX queues on any interface. */
+	uint16_t gatekeeper_num_rx_queues = net_conf->front.num_rx_queues +
+		(net_conf->back_iface_enabled ?
+		 net_conf->back.num_rx_queues : 0);
+
+	/* The total number of TX queues on any interface. */
+	uint16_t gatekeeper_num_tx_queues = net_conf->front.num_tx_queues +
+		(net_conf->back_iface_enabled ?
+		 net_conf->back.num_tx_queues : 0);
+
+	const char *filtered_blocks[] = { "dynamic_conf" };
+
+	/*
+	 * The total number of lcores used by any functional block
+	 * for either RX or TX except for Dynamic Configuration.
+	 */
+	int gatekeeper_num_lcores = launch_count_lcores(filtered_blocks,
+		RTE_DIM(filtered_blocks));
+
+	/*
+	 * The maximum number of receive descriptors to
+	 * allocate for the receive ring over all interfaces.
+	 */
+	uint16_t gatekeeper_num_rx_desc =
+		net_conf->front.gatekeeper_num_rx_desc;
+
+	/*
+	 * The maximum number of transmit descriptors to
+	 * allocate for the transmit ring over all interfaces.
+	 */
+	uint16_t gatekeeper_num_tx_desc =
+		net_conf->front.gatekeeper_num_tx_desc;
+
+	if (net_conf->back_iface_enabled) {
+		gatekeeper_num_rx_desc = RTE_MAX(gatekeeper_num_rx_desc,
+			net_conf->back.gatekeeper_num_rx_desc);
+
+		gatekeeper_num_tx_desc = RTE_MAX(gatekeeper_num_tx_desc,
+			net_conf->back.gatekeeper_num_tx_desc);
+	}
+
+	/*
+	 * The number of elements in the mbuf pool.
+	 *
+	 * The optimum size (in terms of memory usage) for a mempool is when
+	 * it is a power of two minus one.
+	 *
+	 * Need to provision enough memory for the worst case. According to
+	 * l3fwd sample application, it's essentially the number of RX
+	 * descriptors (across all queues for all ports), the number of TX
+	 * descriptors (across all queues for all ports), the number of packet
+	 * burst buffers (across all lcores for all bursts), the number of
+	 * slots in the packet buffer cache (across all lcores).
+	 */
+	net_conf->gatekeeper_num_mbuf = rte_align32pow2(
```

Comment:
1. This formula is not accounting for packets needed in the KNI. Once the problem is solved, remember to update the comment above. @cjdoucette, could you help with this?

2. The value passed to `rte_align32pow2()` is rather important. Define the local variable `gatekeeper_max_num_pkt` to receive this value and log this new local variable.

Reply:
1. Notice that it may be possible to account for the KNI directly in `run_cps()` using the global parameters `net_conf->gatekeeper_total_front_pkt_burst` and `net_conf->gatekeeper_total_back_pkt_burst` simillarly to what I did in my comment below to account for the queue of the block SOL.